### DISCUSSION TITLE
Override dag run generate_run_id

### DISCUSSION THREAD

**Discussion by NilsJPWerner** [apache/airflow#24551] [General]
### Description

 #21851 Added the ability to override run_ids in the manual dag conf. It would be useful to specify a default function for a specific dag to override the dag [generate_run_id](https://github.com/apache/airflow/blob/main/airflow/models/dagrun.py#L407) method for scheduled runs, and allow custom run id formats.

### Use case/motivation

An example use case: You are using the dag run id to determine where files should be saved for a certain run on s3 and you want to add more information to the run_id or use integer timestamps and have that map to the dropdown run ids in the airflow ui.

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by potiuk** [comment]
Why not simply specify the run_id when you trigger the dag via API? you can specify run_id there as you wish.

**Reply by NilsJPWerner** [reply]
That wouldn’t solve for wanting custom run ids for scheduled dag runs. Right now you can have a custom run if for runs triggered in the ui or via the api but not for scheduled runs.

**Reply by potiuk** [reply]
You mentioned manual dag run conf above? Why do you want to convert scheduled runs run id ? what purpose would it serve ?

**Reply by NilsJPWerner** [reply]
I was just saying that the pr I mentioned added that ability and that it would be great to extend custom ids to scheduled runs. It would be great to be able to use custom ids in s3 and other tools that match the ids in the airflow ui.

**Reply by potiuk** [reply]
Why not use tags?

**Reply by NilsJPWerner** [reply]
Isn't that dag level? That wouldn't impact the run ids in this dropdown right?
<img width="484" alt="Screen Shot 2022-06-20 at 2 56 04 PM" src="https://user-images.githubusercontent.com/10968348/174663441-9111cc71-ea03-4402-8c6c-36a86c9ebd32.png">

**Reply by potiuk** [reply]
No. It would not. Question is why can't you do it in other ways - for example by creating an operator extra link to point to the right file on S3 for exampla? This is where  you have control and can define the links:

https://airflow.apache.org/docs/apache-airflow/stable/howto/define_extra_link.html

Why to link those on DAG run level? Imho they should be linked at task instance level.

**Reply by NilsJPWerner** [reply]
I think it’s a common desire to have a universal identifier across services to quickly identify corresponding data or logs connected to a run. Using an extra link can help find data on another platform going one way, but it doesn’t help going the other way where you might want to find the airflow run that corresponds to the data you are looking at on for example Neptune or a trained tensorflow model. 

For example I have a use case where I train a few models in the same dag across a bunch of tasks. They get pushed and deployed to other systems. I use the integer Unix time stamp of the execution date to identify the set of models and the training data because the hosting platform doesn’t support hyphens and dashes. Now when I want to see the training logs of a specific model run I can’t easily find the specific dag run from the top level ui.

**Reply by NilsJPWerner** [reply]
Since airflow does support custom run ids it doesn’t seem like a huge leap to support them in more than just runs that are manually triggered.

**Reply by potiuk** [reply]
The big issue with that proposal is security. By definition airflow scheduler does not execute code provided by the dag author. And we cannot and will not change that. The only code executed by airflow.when scheduling a dag run might only come from installed custom timetable plugins. 

This limitation does not apply to manually or API triggered runs because the custom run id comes from outside - it is not executed by scheduler in the form of executable Python code. 

With scheduled runs - there is no user interaction and your proposal would essentially mean that scheduler will have to execute the code that is provided by DAG author. This is not even possible today because scheduler does not even access the code - of DAG to schedule DAG runs - it only accesses.thr json-serialized form of the DAG in the database, which does not have a code to execute (and it will not have). 

You could potentially try to think about a 'dag run pluginnable function' but that would require to install a custom plugin similar to custom timetables, but that would heavily undermine the whole idea because authors would have to ask admins to install such plugns before they could use it. And for that - better idea would be a way to customise dag run id in the timetable plugins rather andding new entity - though it would not be possible to add 'per external service' approach (for example binding it with Neptune or S3).

I think the only way you could think of it is if you add a metadata for task instance that could be calculated at the moment of task execution - this is the only place where you can safely execute code coming from dag author (it can also be executed at dag parsing time but since you are talking about different values per dag run, dag parsing is not good place.

And at that time run IDs are already defined and immutable. So using it for your use case is out of the question.

Rather than having custom id per dag run - you really need to think about 'per task instance metadata' - similar to cuatom opretor links but stored in the database. You would have to make a proposal on how to add such custom metadata ids generated ad task execution time that would allow you to query for the taks instance (and allow you to find the right dag run).

But usefuleness and complexity of that are rather questionable when you look at complexity/benefit ratio. And actually when you think of it - it's not necessary - you could even today use  XCOM for that and store 'per taks instance metadata' in XCOM by the first task of your dag run. 

I think if you want to keep 'id-matching' between the different systems a bit better solution might be to reverse it. Rather than customise the run id for scheduled tasks, you should make sure that your external systems (Neptune, S3) store the run id of such airflow dag run (for example you could name your S3 file using airflow dag run id or you could add 'airflow_run_id' as label/metadata in you Neptune entity.

But i also think this approach is extremely limited and you will not escape from the reality that each of such systems has its own id and rather than trying to 'bend' one of the systems and 'force' the system id to match another system id, better is to bind the IDs together through common metadata store. Otherwise you are heavily limiting yourself - what happens if you add another system where you also will not be able to change the id (as you cannot change it with Airflow?)

The reality is that you simply need to keep extra metadata to bind such separate systems together. 

And when you think of it actually what you are looking at doing is to build a poor's man lineage system where you could track provenance of data between different systems. What I'd encourage you to do instead of trying to come up with heavily limited and custom approach is to integrate this systems using Open Lineage. Airflow as of 2.3  has first-class support for Open Lineage integration - you can take a look at the few talks a out open lineage from the Airflow Summit 2022 which finished 3 weeks ago. You will easily find relevant talks  by looking at observability chapter of my 'Airflow Summit 2022 The Best of' - blog post: https://link.medium.com/mup3l8F51qb - open lineage provides APiS and there is even reference Marquez implementation that allows you to efficiently link and query the metadata.

I highly recommend you to stand on the shoulders of guants and rather than invent poor's version of lineage, use what is already becoming s de facto standard.

**Comment by potiuk** [comment]
Converting it into a discussion until this is clarified.
